exotica_core/src/scene.cpp

```diff
@@ -603,25 +603,31 @@ void Scene::UpdateSceneFrames()
             Eigen::Isometry3d obj_transform;
             obj_transform.translation() = object.second->shape_poses_[0].translation();
             obj_transform.linear() = object.second->shape_poses_[0].rotation();
-            kinematica_.AddEnvironmentElement(object.first, obj_transform);
-
+            std::shared_ptr<KinematicElement> element = kinematica_.AddEnvironmentElement(object.first, obj_transform);
+            std::vector<VisualElement> visuals;
             for (int i = 0; i < object.second->shape_poses_.size(); ++i)
             {
                 Eigen::Isometry3d shape_transform;
                 shape_transform.translation() = object.second->shape_poses_[i].translation();
                 shape_transform.linear() = object.second->shape_poses_[i].rotation();
                 Eigen::Isometry3d trans = obj_transform.inverse() * shape_transform;
+                VisualElement visual;
+                visual.name = object.first;
+                visual.shape = shapes::ShapePtr(object.second->shapes_[i]->clone());
+                tf::transformEigenToKDL(trans, visual.frame);
                 if (ps_->hasObjectColor(object.first))
                 {
                     auto color_msg = ps_->getObjectColor(object.first);
-                    Eigen::Vector4d color = Eigen::Vector4d(color_msg.r, color_msg.g, color_msg.b, color_msg.a);
-                    kinematica_.AddEnvironmentElement(object.first + "_collision_" + std::to_string(i), trans, object.first, object.second->shapes_[i], KDL::RigidBodyInertia::Zero(), color);
+                    visual.color = Eigen::Vector4d(color_msg.r, color_msg.g, color_msg.b, color_msg.a);
+                    kinematica_.AddEnvironmentElement(object.first + "_collision_" + std::to_string(i), trans, object.first, object.second->shapes_[i], KDL::RigidBodyInertia::Zero(), visual.color);
                 }
                 else
                 {
                     kinematica_.AddEnvironmentElement(object.first + "_collision_" + std::to_string(i), trans, object.first, object.second->shapes_[i]);
                 }
+                if (visual.shape) visuals.push_back(visual);
```

Comment:
reserve, emplace_back

Reply:
Can't reserve because the size gets decided based on `if (visual.shape)`.